libraries/standard/http/include/http_client.h

```diff
@@ -0,0 +1,339 @@
+     * @param[in] statusCode The HTTP response status-code.
+     */
+    void (*onHeaderCallback)( void* pContext, const char* fieldLoc, size_t fieldLen, const char* valueLoc, size_t valueLen, uint16_t statusCode );
+
+    /* Private context for the application. */
+    void *pContext;
+} httpclientHeaderParsingCallback_t;
+
+typedef struct HTTPResponse
+{
+    /**
+     * @brief Buffer for both the raw HTTP header and body. 
+     * 
+     * This buffer cannot be modifed until #HTTPClient_Send returns.
+     */
+    uint8_t* pBuffer;
+    size_t bufferLen; /**< The length of the response buffer in bytes. */
+
+    /**
+     * @brief Optional callback for intercepting the header during the first 
+     * parse through of the response as is it receive from the network.
+     */
+    httpclientHeaderParsingCallback_t* pHeaderParsingCallback;
+    
+    /**
+     * @brief The Location of the response headers in pBuffer.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    uint8_t* pHeaders;
+
+    /**
+     * @brief Byte length of the response headers in pBuf.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    size_t headersLen; 
+
+    /**
+     * @brief The location of the response body in pBuf.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    uint8_t* pBody;
+
+    /**
+     * @brief Byte length of the body in pBuf.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    size_t bodyLen;
+    
+    /* Useful HTTP header values found. */
+
+    /**
+     * @brief The HTTP response Status-Code.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    uint16_t statusCode;
+    
+    /**
+     * @brief The value in the "Content-Length" header is returned here.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    size_t contentLength;
+
+    /**
+     * @brief Count of the headers sent by the server.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    size_t headerCount;
+
+    /**
+     * @brief Flags of useful headers found in the response.
+     * 
+     * This is updated by #HTTPClient_Send.
+     */
+    uint32_t flags;
+} HTTPResponse_t;
+
+
+/**
+ * TODO: Add through documentation.
+ */
+HTTPStatus_t HTTPClient_InitializeRequestHeaders( HTTPRequestHeaders_t * pRequestHeaders, 
+                                                      const HTTPRequestInfo_t * pRequestInfo );
+
+/**
+ * TODO: Add thorough documentation.
+ */
+HTTPStatus_t HTTPClient_AddHeader( HTTPRequestHeaders_t * pRequestHeaders, 
+                                       const char* pName, 
+                                       size_t nameLen, 
+                                       const char * pValue, 
+                                       size_t valueLen );
+
+/**
+ * TODO: Add thorough documentation.
+ */
+HTTPStatus_t HTTPClient_AddRangeHeader( HTTPRequestHeaders_t *pRequestHeaders,
+                                            int32_t rangeStart,
```

Comment:
Why are these of _signed_ integer types?

Reply:
rangeStart and rangeEnd can be const but all the same since it's passing by copy

Reply:
Read the design document to find out why these are signed.

Reply:
https://w.amazon.com/bin/view/AWS/Mobile/IoT/Treadstone/HTTPDesign/Redesign/LowLevel/#HHTTPRangeRequestSupport

I could change rangeEnd to uint32_t.

Reply:
I will leave it up to the person implementing this.